File: ch8/8-1.c
#include <stdio.h>
#include <unistd.h>
#include <fcntl.h>
#include <stdarg.h>
#include <stdlib.h>

void error(char *, ...);
void filecopy(int from, int to);

int main(int argc, char *argv[])
{
    int f;
    if (argc == 1) {
        filecopy(0, 1);
    } else {
        while (--argc > 0) {
            if ((f = open(*++argv, O_RDONLY)) == -1) {
                error("can't open %s", *argv);
                return 1;
            } else {
                filecopy(f, 1);
                close(f);
            }
        }
    }
    return 0;
}

void error(char * fmt, ...)
{
    va_list args;

    va_start(args, fmt);
    fprintf(stderr, "error: ");
    vfprintf(stderr, fmt, args);
    fprintf(stderr, "\n");
    va_end(args);
    exit(1);
}

void filecopy(int from, int to)
{
    int n;
    char buf[BUFSIZ];
    while ((n = read(from, buf, BUFSIZ)) > 0) {
        write(to, buf, n);
    }
}

File: ch8/8-1.go
package main

import (
	"fmt"
	"io"
	"os"
)

func main() {
	if len(os.Args) == 1 {
		filecopy(os.Stdin, os.Stdout)
		return
	}
	for i, f := range os.Args {
		if i == 0 {
			continue
		}
		file, err := os.Open(f)
		if err != nil {
			fmt.Printf("%s\n", err)
			os.Exit(1)
		}
		filecopy(file, os.Stdout)
		err = file.Close()
		if err != nil {
			fmt.Printf("%s\n", err)
			os.Exit(1)
		}
	}
}

func filecopy(from io.Reader, to io.Writer) {
	_, err := io.Copy(to, from)
	if err != nil {
		fmt.Printf("%s\n", err)
		os.Exit(1)
	}
}
